### PR TITLE
WIP: initial commit for a predict_proba with threshold

### PR DESCRIPTION
**Pull request by Valleo4**
- [x] closes issue #463 
- [x] Tests added
- [x] Passes flake8
- [ ] Updated CHANGELOG
- [ ] Updated README.md

### CODE REVIEW

**Comment by Valleo4** [comment]
@andersbogsnes can you take a look here? I think this solution only works for binary classifiers though, any ideas?

File: src/ml_tooling/result/viz/classification_viz.py [hunk 1]
```diff
@@ -15,7 +16,9 @@ class ClassificationVisualize(BaseVisualize):
     Visualization class for Classification models
     """
 
-    def confusion_matrix(self, normalized: bool = True, **kwargs) -> plt.Axes:
+    def confusion_matrix(
+        self, normalized: bool = True, threshold: float = None, **kwargs
```
Comment (andersbogsnes) [inline src/ml_tooling/result/viz/classification_viz.py]: We can set default value to 0.5

File: src/ml_tooling/result/viz/classification_viz.py [hunk 2]
```diff
@@ -33,7 +36,15 @@ def confusion_matrix(self, normalized: bool = True, **kwargs) -> plt.Axes:
 
         with plt.style.context(self._config.STYLE_SHEET):
             title = f"Confusion Matrix - {self._estimator_name}"
-            y_pred = self._estimator.predict(self._data.test_x)
+            if threshold is None:
```
Comment (andersbogsnes) [inline src/ml_tooling/result/viz/classification_viz.py]: With default value, we can get rid of the conditional

File: src/ml_tooling/result/viz/classification_viz.py [hunk 3]
```diff
@@ -33,7 +36,15 @@ def confusion_matrix(self, normalized: bool = True, **kwargs) -> plt.Axes:
...
+                y_pred = self._estimator.predict(self._data.test_x)
+            else:
+                y_prob = self._estimator.predict_proba(self._data.test_x)
```
Comment (andersbogsnes) [inline src/ml_tooling/result/viz/classification_viz.py]: I think this does the same - should maybe be `preds >= threshold`? Check sklearn implementation?

```
y_prob = self._estimator.predict_proba(self._data.test_x)
y_pred = np.where((preds > threshold) & (preds == preds.max(axis=1, keepdims=True)), 1, 0)
```

File: src/ml_tooling/result/viz/classification_viz.py (hunk 2 above)
Comment (andersbogsnes) [reply src/ml_tooling/result/viz/classification_viz.py]: Are there any classifiers which don't implement `predict_proba`?

File: src/ml_tooling/result/viz/classification_viz.py [hunk 4]
```diff
@@ -15,7 +16,9 @@ class ClassificationVisualize(BaseVisualize):
     Visualization class for Classification models
     """
 
-    def confusion_matrix(self, normalized: bool = True, **kwargs) -> plt.Axes:
+    def confusion_matrix(
+        self, normalized: bool = True, threshold: float = None, **kwargs
+    ) -> plt.Axes:
         """
         Visualize a confusion matrix for a classification estimator
```
Comment (andersbogsnes) [inline src/ml_tooling/result/viz/classification_viz.py]: Missing threshold in docstring

File: tests/test_visualize.py [hunk 5]
```diff
@@ -81,6 +81,19 @@ def test_confusion_matrix_plots_have_correct_data(self, classifier: Model):
         assert "Predicted Label" == ax.get_xlabel()
         plt.close()
 
+    def test_confusion_matrix_plots_have_correct_data_with_threshold(
```
Comment (andersbogsnes) [inline tests/test_visualize.py]: Maybe add a test for the `np.where` line? It's easier to parametrize it and pass edge cases

File: src/ml_tooling/result/viz/classification_viz.py (hunk 2 above)
Comment (Valleo4) [reply src/ml_tooling/result/viz/classification_viz.py]: Dont think so, im thinking more of multiclass classifiers, the shape of predict_proba depends on the number of classes (.classes_)

Comment (andersbogsnes) [reply src/ml_tooling/result/viz/classification_viz.py]: Then it’s safe to always use `pred_proba`

File: src/ml_tooling/result/viz/classification_viz.py (hunk 3 above)
Comment (Valleo4) [reply src/ml_tooling/result/viz/classification_viz.py]: Seem rights, ill have to make tests as you suggest though. But this certainly assumes that the 0 class is indeed the false class and 1 the true, which i dont know is always the case? That was the reason for the last line in my suggestion

Comment (andersbogsnes) [reply src/ml_tooling/result/viz/classification_viz.py]: Yes, you're right - need to add the indexing part:

`self._estimator.classes_[np.argmax(y_pred, axis=1)]`

File: src/ml_tooling/baseclass.py [hunk 6]
```diff
@@ -266,10 +268,21 @@ def make_prediction(
...
                 data = self.estimator.predict_proba(x)
-                columns = [f"Probability Class {col}" for col in range(data.shape[1])]
+                columns = [
+                    f"Probability Class {col}" for col in self.estimator.classes_
+                ]
             else:
                 data = self.estimator.predict(x)
                 columns = ["Prediction"]
+            if threshold is not None:
+                y_prob = data if proba else self.estimator.predict_proba(x)
```
Comment (andersbogsnes) [inline src/ml_tooling/baseclass.py]: How does this code work if we use `data` - this would be a 1D array? Does it work correctly if we set threshold to 0.5? and run the threshold code against `np.array([1, 2, 1, 0])`

Comment (andersbogsnes) [reply src/ml_tooling/baseclass.py]: Then we could go back to default 0.5 if that works correctly

File: src/ml_tooling/baseclass.py [hunk 7]
```diff
@@ -226,6 +227,7 @@ def make_prediction(
         data: Dataset,
         *args,
         proba: bool = False,
```
Comment (andersbogsnes) [inline src/ml_tooling/baseclass.py]: Alternatively, proba could be a float and that's the threshold, if you can't have both `proba=False` and `threshold=0.2` - would probably have to rename it though

File: src/ml_tooling/baseclass.py [hunk 8]
```diff
@@ -266,10 +268,21 @@ def make_prediction(
         try:
             if proba:
                 data = self.estimator.predict_proba(x)
-                columns = [f"Probability Class {col}" for col in range(data.shape[1])]
+                columns = [
+                    f"Probability Class {col}" for col in self.estimator.classes_
```
Comment (andersbogsnes) [inline src/ml_tooling/baseclass.py]: :+1:

File: src/ml_tooling/result/viz/classification_viz.py [hunk 9]
```diff
@@ -33,7 +38,13 @@ def confusion_matrix(self, normalized: bool = True, **kwargs) -> plt.Axes:
 
         with plt.style.context(self._config.STYLE_SHEET):
             title = f"Confusion Matrix - {self._estimator_name}"
-            y_pred = self._estimator.predict(self._data.test_x)
+            y_prob = self._estimator.predict_proba(self._data.test_x)
```
Comment (andersbogsnes) [inline src/ml_tooling/result/viz/classification_viz.py]: Maybe a utility function to handle this? Easier to write tests for

File: tests/test_baseclass.py [hunk 10]
```diff
@@ -726,6 +726,25 @@ def test_make_prediction_returns_proba_if_proba_is_true(
             start=expected_index, stop=expected_index + 1, step=1
         )
 
+    @pytest.mark.parametrize("use_index, expected_index", [(False, 0), (True, 5)])
+    def test_make_prediction_returns_prediction_if_threshold_is_specified(
+        self,
+        classifier: Model,
+        use_index: bool,
+        expected_index: int,
+        train_iris_dataset,
```
Comment (andersbogsnes) [inline tests/test_baseclass.py]:
```suggestion
        train_iris_dataset: Dataset,
```

File: src/ml_tooling/result/viz/classification_viz.py (hunk 2 above)
Comment (andersbogsnes) [reply src/ml_tooling/result/viz/classification_viz.py]: Was the result that not all classifiers have a predict_proba?

Comment (Valleo4) [reply src/ml_tooling/result/viz/classification_viz.py]: Yes, not all classifiers have a predict_proba unfortunately, so i tried working around that

File: src/ml_tooling/baseclass.py (hunk 6 above)
Comment (Valleo4) [reply src/ml_tooling/baseclass.py]: It works correctly with 0.5 threshold i think, the question is if we can use data, ill have to check that out

File: src/ml_tooling/baseclass.py (hunk 7 above)
Comment (Valleo4) [reply src/ml_tooling/baseclass.py]: I'll look into it

File: src/ml_tooling/baseclass.py (hunk 6 above)
Comment (Valleo4) [reply src/ml_tooling/baseclass.py]: This wasnt possible without transforming the 1d array into the shape of predict_proba

File: tests/test_utils.py [hunk 11]
```diff
@@ -176,3 +177,15 @@ def test_find_src_dir_errors_when_no_init_is_found(self, tmp_path: pathlib.Path)
...
+    def test_classify_with_threshold_works_on_classification_output(
+        self, classifier, train_iris_dataset
+    ):
+        data = train_iris_dataset.load_prediction_data(0)
+
+        pred = _classify(data, classifier.estimator)
+        assert pred == np.array([0])
+
+        pred = _classify(data, classifier.estimator, threshold=0.20)
+        # this should be true since the true proba exceeds the threshold?
```
Comment (Valleo4) [inline tests/test_utils.py]: We need to pick the true classification when it exceeds the threshold, right? Right now the false is result if both exceeds the threshold

Comment (andersbogsnes) [reply tests/test_utils.py]: Hmm, you're right - I think in this round we should focus on getting the binary situation handled - I'm not sure thresholding makes sense for multiple classes...

File: src/ml_tooling/utils.py [hunk 12]
```diff
@@ -362,3 +362,30 @@ def _find_src_dir(path: pathlib.Path = None, max_level: int = 2) -> pathlib.Path
...
+        The estimator to use for making the prediction
+    threshold: float
+        Threshold of classification
+
+    Returns
+    -------
+    np.array
+        Array of class predictions
+    """
+    y_prob = estimator.predict_proba(x)
```
Comment (andersbogsnes) [inline src/ml_tooling/utils.py]: If we have a check for the binary classification (something like `len(estimator.classes_) == 2`) we can do a simple comparison: `return (y_pred[:, 1] > threshold).astype(np.int32)`

In binary classification 0 will always be negative class and 1 positive class

File: tests/test_baseclass.py [hunk 13]
```diff
@@ -726,6 +726,25 @@ def test_make_prediction_returns_proba_if_proba_is_true(
...
+        self,
+        classifier: Model,
+        use_index: bool,
+        expected_index: int,
+        train_iris_dataset: Dataset,
+    ):
+        results = classifier.make_prediction(
+            train_iris_dataset, 5, threshold=0.6, use_index=use_index
+        )
+        assert isinstance(results, pd.DataFrame)
```
[the remark this answers is not in the document]
Comment (Valleo4) [reply tests/test_baseclass.py]: This isnt what the previous tests do, so i just did the same :)

File: src/ml_tooling/utils.py (hunk 12 above)
Comment (Valleo4) [reply src/ml_tooling/utils.py]: Yeah, this is way simpler :)
